test/utils/wasm3_engine.cpp

```diff
@@ -0,0 +1,105 @@
+#include "source/wasm3.h"
+#include <test/utils/wasm_engine.hpp>
+#include <cassert>
+#include <cstring>
+
+namespace fizzy::test
+{
+class Wasm3Engine : public WasmEngine
+{
+    IM3Environment m_env{nullptr};
+    IM3Runtime m_runtime{nullptr};
+
+public:
+    Wasm3Engine() : m_env{m3_NewEnvironment()} {}
+    ~Wasm3Engine()
+    {
+        if (m_runtime)
+            m3_FreeRuntime(m_runtime);
+        m3_FreeEnvironment(m_env);
+    }
+
+    bool parse(bytes_view input) final;
+    std::optional<FuncRef> find_function(std::string_view name) const final;
+    bool instantiate() final;
+    fizzy::bytes_view get_memory() const final;
+    void set_memory(fizzy::bytes_view memory) final;
+    Result execute(FuncRef func_ref, const std::vector<uint64_t>& args) final;
+};
+
+std::unique_ptr<WasmEngine> create_wasm3_engine()
+{
+    return std::make_unique<Wasm3Engine>();
+}
+
+bool Wasm3Engine::parse(bytes_view input)
+{
+    // Replace runtime (e.g. instance + module)
+    if (m_runtime != nullptr)
+        m3_FreeRuntime(m_runtime);
+
+    // The 64k stack size comes from wasm3/platorms/app
+    m_runtime = m3_NewRuntime(m_env, 64 * 1024, nullptr);
+    if (m_runtime == nullptr)
+        return false;
+
+    IM3Module module;
+    if (m3_ParseModule(m_env, &module, input.data(), uint32_t(input.size())) != m3Err_none)
+        return false;
+
+    // Transfers ownership to runtime.
+    if (m3_LoadModule(m_runtime, module) != m3Err_none)
```

Comment:
Don't creating runtime and loading module belong more to `instantiate`?

Reply:
See https://github.com/wasmx/fizzy/pull/113#discussion_r375894251

But for some reason it is crashing. Likely some ownership issue internally.